contribs/gmf/src/directives/layertree.js

```diff
@@ -175,6 +176,46 @@ gmf.LayertreeController = function($http, $sce, $scope, ngeoCreatePopup,
    */
   this.dataLayerGroup_ = this.layerHelper_.getGroupFromMap(this.map,
         gmf.DATALAYERGROUP_NAME);
+
+  $scope.$watchCollection('gmfLayertreeCtrl.tree.children', function(newNodes, oldNodes) {
+    var ctrl = $scope['gmfLayertreeCtrl'];
+    var reorderedLayers = ctrl.reorderLayer_(newNodes, oldNodes, ctrl.dataLayerGroup_.getLayers());
+    if (reorderedLayers) {
+      ctrl.dataLayerGroup_.setLayers(reorderedLayers);
+    }
+  });
+};
+
+/**
+ * Reorder a layer collection regarding changes in the related nodes
+ *
+ * (1) : if newNodes and oldNodes are equals, return null (there is no need to reorder)
+ * (2) : newNodes and oldNodes MUST HAVE the same length (avoiding reordering on node insertion/deletion)
+ * (3) : newNodes (or oldNodes) and layers MUST HAVE the same length
+ *
+ * @param {Array<Object>} newNodes nodes from the tree
+ * @param {Array<Object>} oldNodes nodes from the tree
+ * @param {ol.Collection} layers a collection of layers
+ * @return {ol.Collection|null} a new (reordered) layer collection
+ * @private
+ */
+gmf.LayertreeController.prototype.reorderLayer_ = function(newNodes, oldNodes, layers) {
+  var sortedLayers, prevIndex, layer;
+  if (newNodes && newNodes.length === oldNodes.length && newNodes !==  oldNodes && newNodes.length === layers.getLength()) {
```

Comment:
It doesn't cover 100% of the cases... but might be enough for now.
Note that if a sub layer is  a WMTS, needing a xhr, and you change order before you get response, you're screwed.